crates/red_knot_module_resolver/src/testing.rs

```diff
@@ -0,0 +1,303 @@
+    /// Specify files to be created in the `src` mock directory
+    pub(crate) fn with_src_files(mut self, files: &[FileSpec]) -> Self {
+        self.first_party_files.extend(files.iter().copied());
+        self
+    }
+
+    /// Specify files to be created in the `site-packages` mock directory
+    pub(crate) fn with_site_packages_files(mut self, files: &[FileSpec]) -> Self {
+        self.site_packages_files.extend(files.iter().copied());
+        self
+    }
+
+    /// Specify the target Python version the module resolver should assume
+    pub(crate) fn with_target_version(mut self, target_version: TargetVersion) -> Self {
+        self.target_version = target_version;
+        self
+    }
+
+    fn write_mock_directory(
+        location: impl AsRef<SystemPath>,
+        system: &MemoryFileSystem,
+        files: impl IntoIterator<Item = FileSpec>,
+    ) -> SystemPathBuf {
+        let root = location.as_ref().to_path_buf();
+        system.create_directory_all(&root).unwrap();
+        for (path, contents) in files {
+            let path = root.join(path);
+            if let Some(parent) = path.parent() {
+                system.create_directory_all(parent).unwrap();
+            }
+            system.write_file(path, contents).unwrap();
+        }
+        root
+    }
+}
+
+impl TestCaseBuilder<UnspecifiedTypeshed> {
+    pub(crate) fn new() -> TestCaseBuilder<UnspecifiedTypeshed> {
+        Self {
+            typeshed_option: UnspecifiedTypeshed,
+            target_version: TargetVersion::default(),
+            first_party_files: vec![],
+            site_packages_files: vec![],
+        }
+    }
+
+    /// Use the vendored stdlib stubs included in the Ruff binary for this test case
+    pub(crate) fn with_vendored_typeshed(self) -> TestCaseBuilder<VendoredTypeshed> {
+        let TestCaseBuilder {
+            typeshed_option: _,
+            target_version,
+            first_party_files,
+            site_packages_files,
+        } = self;
+        TestCaseBuilder {
+            typeshed_option: VendoredTypeshed,
+            target_version,
+            first_party_files,
+            site_packages_files,
+        }
+    }
+
+    /// Use a mock typeshed directory for this test case
+    pub(crate) fn with_custom_typeshed(
+        self,
+        typeshed: MockedTypeshed,
+    ) -> TestCaseBuilder<MockedTypeshed> {
+        let TestCaseBuilder {
+            typeshed_option: _,
+            target_version,
+            first_party_files,
+            site_packages_files,
+        } = self;
+        TestCaseBuilder {
+            typeshed_option: typeshed,
+            target_version,
+            first_party_files,
+            site_packages_files,
+        }
+    }
+
+    pub(crate) fn build(self) -> TestCase<()> {
+        const DEFAULT_TYPESHED: MockedTypeshed = MockedTypeshed {
+            stdlib_files: &[("functools.pyi", "def update_wrapper(): ...")],
+            versions: "functools: 3.8-",
```

Comment:
This seems a bit arbitrary?

Reply:
Yeah, it _is_ arbitrary. However, we do need _something_ in the `VERSIONS` file or it will fail to parse:

https://github.com/astral-sh/ruff/blob/88abc6aed801995864bb4f8bcb401d1c2cb48baf/crates/red_knot_module_resolver/src/typeshed/versions.rs#L310-L314

E.g. two tests that don't test anything to do with the stdlib start failing if I make this diff:

```diff
diff --git a/crates/red_knot_module_resolver/src/testing.rs b/crates/red_knot_module_resolver/src/testing.rs
index 55f0d9d0a..ce38aca5c 100644
--- a/crates/red_knot_module_resolver/src/testing.rs
+++ b/crates/red_knot_module_resolver/src/testing.rs
@@ -186,7 +186,7 @@ impl TestCaseBuilder<UnspecifiedTypeshed> {
     pub(crate) fn build(self) -> TestCase<()> {
         const DEFAULT_TYPESHED: MockedTypeshed = MockedTypeshed {
             stdlib_files: &[("functools.pyi", "def update_wrapper(): ...")],
-            versions: "functools: 3.8-",
+            versions: "",
         };
 
         let TestCase {
```

And if the `VERSIONS` file says a module exists, then we probably should indeed put that module in the mock typeshed...

Maybe I should just put an empty `foo.pyi` file in the default typeshed mock, and put `foo` in the `VERSIONS` file? And add a comment? WDYT?

Reply:
My initial feeling is that a VERSIONS file that exists but is empty should not be an error in the parser, it just means there are no stdlib stub files. And that should be the default for tests that don't either provide a mock typeshed or request to use the vendored typeshed.

Reply:
Yeah, that makes sense. Two counterpoints, however:
- I think it is useful to validate _somewhere_ that the contents of a custom typeshed directory a user passes to us makes some kind of sense. For example, mypy checks that a small selection of "core" stdlib modules exist in typeshed, and aborts the process immediately if they don't, since it knows no typechecking it does is going to make any kind of sense without them (`builtins`, `typing`, `collections.abc`, `types`, `typing_extensions`, I believe). I think it makes sense for us to do similarly at some place in `red_knot` (but it doesn't necessarily have to be in the `VERSIONS` parser).
- I'm somewhat uneasy about the fact that `File::read_to_string()` just returns an empty string if a file doesn't exist... in the specific ways we're currently using the `ruff_db` API to get the `VERSIONS` file in `versions.rs`, it should be impossible for us to try parsing an empty string that represents a nonexistent `VERSIONS` file, but I still sort of worry about it... https://github.com/astral-sh/ruff/blob/88abc6aed801995864bb4f8bcb401d1c2cb48baf/crates/ruff_db/src/files.rs#L181-L190

Reply:
Regarding Files::read_to_string. The fallback path is only taken if the file has been deleted since the last system_path_to_file call. The analysis will most likely be aborted shortly after once the file watcher detects the change and marks the file as deleted (and system_path_to_fike returns None)

Reply:
I guess I am torn here between a) do the simple and obvious thing, and don't unnecessarily prevent users from doing what they want, even if it's odd (for example, an empty typeshed), and b) prevent users from shooting themselves in the foot accidentally.

I think that we should be able to roll with an empty stdlib, even an empty builtins; it'll just result in a lot of Unknown/Unbound in our type inference.

I guess I feel like we should either allow empty VERSIONS, or require an actually-usable stdlib (at least has builtins). Requiring a non-empty-but-can-have-whatever-nonsense-you-want-in-it isn't a middle ground that seems very useful.

Reply:
Alrighty. For now I'll allow an empty `VERSIONS`, then. We can add more principled validation if/when we feel we need it.

Reply:
> The analysis will most likely be aborted shortly after once the file watcher detects the change and marks the file as deleted (and system_path_to_fike returns None)

Right. That still feels like an assumption that I feel like I'd prefer to check explicitly (and panic if it turns out to be incorrect). But not sure; happy to defer to you.

Reply:
I like to verify that as well. But it should not panic, because that tears down the entire program. Rust analyzer has a few interesting techniques. It keeps the last working workspace tree around for the case that a user messes up the editing of a Cargo.toml. They use the old workspace structure in that case because it's still better than knowing nothing about the program.